src/EditorFeatures/Test/Preview/PreviewWorkspaceTests.cs

```diff
@@ -249,16 +237,8 @@ public void TestPreviewDiagnostic()
                         var rightTagger = rightProvider.CreateTagger<IErrorTag>(rightBuffer);
                         using (var rightDisposable = rightTagger as IDisposable)
                         {
-                            // wait up to 20 seconds for diagnostics
-                            taskSource.Task.Wait(20000);
-                            if (!taskSource.Task.IsCompleted)
-                            {
-                                // something is wrong
-                                FatalError.Report(new System.Exception("not finished after 20 seconds"));
-                            }
-
-                            // wait taggers
-                            await listenerProvider.GetWaiter(FeatureAttribute.ErrorSquiggles).CreateWaitTask();
+                            // wait for diagnostics and taggers
+                            await listenerProvider.WaitAllDispatcherOperationAndTasksAsync(FeatureAttribute.DiagnosticService, FeatureAttribute.ErrorSquiggles);
```

Comment:
listener....WaitAsync accepts multiple waiter but, ones that given together is expected not have dependency each other. for DiagnosticService, SolutionCrawler, ErrorSquiggle has dependency each other. like SolutionCrawler must run and finish first, then DiagnosticService and then ErrorSquiggle. so probably better to have things in order

WaitAll...(SolutionCrawler);
WaitAl...(Diagnostic...)
..

like that

Reply:
We should fix this. The operation should complete when none of the above are running, which even handles provided termination occurs.

Reply:
he he, you are right :)